proposals/rich-content-graphql-api.md

```diff
@@ -0,0 +1,305 @@
+
+```graphql
+type SliderSlide {
+  alt: String!
+  src: String!
+}
+
+type PageBuilder_Slider implements VisualContentNode {
+  role: String!
+  depth: Int
+  preload: [String]
+  descendentRoles: [String]
+  children: [VisualContentNode]
+  html: String
+  autoStart: Boolean
+  slides: [SliderSlide]
+}]]>
+```
+
+A particular instance of `Slider` in response to a query for its ID against the `VisualContentNode` interface would look like:
+
+```json
+{
+  "content": {
+    "role": "Slider",
+    "preload": ["slide1.jpg", "slide2.jpg", "slide3.jpg"],
+    "depth": 0,
+    "descendentRoles": [],
+    "children": [],
+    "html": ""
+  }
+}
+```
+
+The PWA would handle this response:
+
+1.  In parallel:
+
+    1.  Render the `html` content by injecting it into a `RawHTML` component.
+
+    2.  Initiate preload requests for all the URLs in the `preload` field.
+
+    3.  In series:
+
+    4.  Dynamically load and inject the `PageBuilder_Slider` React Component. (The bundle system and service worker will load from their caches if possible.)
+    5.  Run a subsequent query provided by the component:
+
+        ```graphql
+        fragment PageBuilder_Slider_Config on PageBuilder_Slider {
+          autoStart
+          slides {
+            node {
+              alt
+              src
+            }
+          }
+        }
+
+        content {
+          ... on PageBuilder_Slider {
+            ...PageBuilder_Slider_Config
+          }
+        }
+        ```
+
+
+        3.  Re-render the fully hydrated Slider component when that query returns results.
+
+Subsequent requests for any content which lists `PageBuilder_Slider` in its `descendentRoles` will automatically include that fragment and clause in the `content` query. Therefore, subsequent requests will contain the necessary slider configuration on the first pass.
+
+Fields which contain non-PageBuilder content (for example, TinyMCE WYSIWYG-generated or imported content) should be of a type `RawHtml` , instances of which have null `children` and `init` fields, a 0 depth field, and a `role` of `RawHTML` or some other signal value, and raw HTML in the `html` field, which the PWA runtime may sanitize and render. In this way, all content in GraphQL is of the same type.
+
+## Challenges and FAQs
+
+- **Why do more network traffic on the first pageload?**
+
+  - Block implementations can provide placeholder content of arbitrary size and complexity; the only constraint is that this content must be "static", with no JS-driven behaviors. Blocks optimized for use on landing pages should have acceptable fallbacks implemented in the \`html\` field resolver.
+
+* **Why not use GraphQL introspection for this?**
+
+  - Per the GraphQL specification, the introspection query must provide only metadata, and not any data from resolvers. Implementing "partial introspection" via type tag names is a more efficient way of resolving recursive content nodes.
```

Comment:
Introspection will also be disabled in production.

Reply:
I should bring this up in a different issue, but we should at least permit the "[standard introspection query"](https://github.com/graphql/graphql-js/blob/master/src/utilities/introspectionQuery.js) in production, which GraphQL clients use to generate documentation. It should be enough to cache the results of this query, since it will be invariant in production.